src/map/intif.c

```diff
@@ -1495,11 +1497,94 @@ void intif_parse_DeleteHomunculusOk(int fd) {
 		ShowError("Homunculus data delete failure\n");
 }
 
-/**************************************
+/***************************************
+ *
+ * ACHIEVEMENT SYSTEM FUNCTIONS
+ *
+ ***************************************/
+/**
+ * Sends a request to the inter-server to load
+ * and send a character's achievement data.
+ * @packet [out] 0x3098 <char_id>.L
+ * @param sd pointer to map_session_data.
+ */
+void intif_achievements_request(struct map_session_data *sd)
+{
+	nullpo_retv(sd);
 
-QUESTLOG SYSTEM FUNCTIONS
+	if (intif->CheckForCharServer())
+		return;
 
-***************************************/
+	WFIFOHEAD(inter_fd, 6);
+	WFIFOW(inter_fd, 0) = 0x3098;
+	WFIFOL(inter_fd, 2) = sd->status.char_id;
+	WFIFOSET(inter_fd, 6);
+}
+
+/**
+ * Handles reception of achievement data for a character from the inter-server.
+ * @packet [in] 0x3898 <packet_len>.W <char_id>.L <char_achievements[]>.P
+ * @param fd socket descriptor.
+ */
+void intif_parse_achievements_load(int fd)
+{
+	int size = RFIFOW(fd, 2);
+	int char_id = RFIFOL(fd, 4);
+	int payload_count = (size - 8) / sizeof(struct achievement);
+	struct map_session_data *sd = map->charid2sd(char_id);
+	int i = 0;
+
+	VECTOR_ENSURE(sd->achievement, payload_count, 1);
+
+	for (i = 0; i < payload_count; i++) {
+		struct achievement t_ach = { 0 };
+
+		memcpy(&t_ach, RFIFOP(fd, 8 + i * sizeof(struct achievement)), sizeof(struct achievement));
+
+		if (achievement->get(t_ach.id) == NULL) {
+			ShowError("intif_parse_achievements_load: Invalid Achievement %d received from character %d. Ignoring...\n", t_ach.id, char_id);
+			continue;
+		}
+
+		VECTOR_PUSH(sd->achievement, t_ach);
```

Comment:
What if this is called twice? (in case the char server sends it twice for any reason)